commit e5ec1a07fb97a2b0e1a337b0d5d134a7181a2679
Author: Andy Whitcroft <apw@canonical.com>
Date:   Tue Dec 17 11:50:04 2013 +0000

    UBUNTU: SAUCE: suspicious unlocked ->status reading and writing in ipc/sem.c
    
    Signed-off-by: Andy Whitcroft <apw@canonical.com>

diff --git a/ipc/sem.c b/ipc/sem.c
index 5e318c5..51cc9f1 100644
--- a/ipc/sem.c
+++ b/ipc/sem.c
@@ -2008,6 +2008,14 @@ sleep_again:
 	error = get_queue_result(&queue);
 
 	/*
+	 * wake_up_sem_queue_do operates on queue without locking, so we
+	 * need a barrier here to order our read of queue.status and the
+	 * subsequent reuse of queue (queue is on the stack so will be
+	 * most likely reused in the next function call).
+	 */
+	smp_mb();
+
+	/*
 	 * Array removed? If yes, leave without sem_unlock().
 	 */
 	if (IS_ERR(sma)) {
